templates/Makefile.build-harness

```diff
@@ -4,21 +4,60 @@
 #
 
 export SHELL = /bin/bash
-export BUILD_HARNESS_PATH ?= $(shell until [ -d "$(BUILD_HARNESS_PROJECT)" ] || [ "`pwd`" == '/' ]; do cd ..; done; pwd)/$(BUILD_HARNESS_PROJECT)
+export PWD = $(shell pwd)
 export BUILD_HARNESS_ORG ?= cloudposse
 export BUILD_HARNESS_PROJECT ?= build-harness
 export BUILD_HARNESS_DOCKER_IMAGE ?= $(BUILD_HARNESS_ORG)/$(BUILD_HARNESS_PROJECT)
 export BUILD_HARNESS_BRANCH ?= master
+export BUILD_HARNESS_CLONE_URL ?= https://github.com/$(BUILD_HARNESS_ORG)/$(BUILD_HARNESS_PROJECT).git
 # It is kind of expensive to figure out the Docker SHA tag, so we just define the command here, and only call it when needed
 export BUILD_HARNESS_DOCKER_SHA_TAG_CMD := git -C "$(BUILD_HARNESS_PATH)" log -n 1 --format=sha-%h 2>/dev/null || echo latest
 
--include $(BUILD_HARNESS_PATH)/Makefile
+# Resolves BUILD_HARNESS_PATH to BUILD_HARNESS_PATH_LOCAL when BUILD_HARNESS_PATH does not exist
+BUILD_HARNESS_PATH ?= $(shell until [ -d "$(BUILD_HARNESS_PROJECT)" ] || [ "`pwd`" == '/' ]; do cd ..; done; pwd)/$(BUILD_HARNESS_PROJECT)
+BUILD_HARNESS_PATH_LOCAL := $(PWD)/$(BUILD_HARNESS_PROJECT)
+export BUILD_HARNESS_PATH := $(or $(wildcard $(BUILD_HARNESS_PATH)),$(BUILD_HARNESS_PATH_LOCAL))
+
+# Toggles the auto-init feature
+BUILD_HARNESS_AUTO_INIT ?= false
+
+# Macro to clone/install BUILD_HARNESS_PROJECT
+define harness_install
+curl --retry 5 --fail --silent --retry-delay 1 \
+	https://raw.githubusercontent.com/$(BUILD_HARNESS_ORG)/$(BUILD_HARNESS_PROJECT)/$(BUILD_HARNESS_BRANCH)/bin/install.sh | \
+	bash -s "$(BUILD_HARNESS_ORG)" "$(BUILD_HARNESS_PROJECT)" "$(BUILD_HARNESS_BRANCH)"
+endef
+
+# Macro to auto-init the BUILD_HARNESS_PROJECT with the `include` directive
+# Tests if BUILD_HARNESS_PROJECT does not yet exist, or if it does exist but the
+# checkout does not match BUILD_HARNESS_BRANCH
+define harness_auto_init
+if [[ \
+	-f "/.dockerenv" && -f "/$(BUILD_HARNESS_PROJECT)/Makefile" \
+]]; then \
+	echo "[.build-harness]: In $(BUILD_HARNESS_PROJECT) docker container, skipping auto-init" ;\
```

Comment:
```suggestion
	-f /build-harness/Makefile || -f "/$(BUILD_HARNESS_PROJECT)/Makefile"  \
]]; then \
	echo "[.build-harness]: In $(BUILD_HARNESS_PROJECT) docker container, skipping auto-init" ;\
elif [[ \
	grep -q docker /proc/1/cgroup 2>/dev/null \
]]; then \
	echo "[.build-harness]: In unknown docker container, skipping auto-init" ;\
```

Reply:
This does not feel quite right... If using a _different_ docker container, you almost certainly *do* want to auto-init, especially if they set `BUILD_HARNESS_AUTO_INIT=true` which is required right now, because that container will not have the build-harness.

Reply:
For example, using the `python:3` container:

```
$ docker run --rm -ti -v "$PWD:/workdir" -e BUILD_HARNESS_AUTO_INIT=true -w /workdir --entrypoint make python:3 help
Cloning https://github.com/cloudposse/build-harness.git#master...
Cloning into 'build-harness'...
remote: Enumerating objects: 143, done.
remote: Counting objects: 100% (143/143), done.
remote: Compressing objects: 100% (118/118), done.
remote: Total 143 (delta 7), reused 71 (delta 3), pack-reused 0
Receiving objects: 100% (143/143), 85.57 KiB | 551.00 KiB/s, done.
Resolving deltas: 100% (7/7), done.
Available targets:

[.build-harness]: Clone of build-harness is up-to-date, skipping auto-init
  aws/install                         Install aws cli bundle
```

Reply:
This is me being cautious and conservative. Any exiting scripted docker image knows to call `make init` first. I want to limit this enhancement to being a convenience for interactive use and not have scripts rely on it. I don't know why you would want to run this in a `python:3` container, but if you do, you still have 2 options:

```
$ make init && \
  docker run --rm -ti -v "$PWD:/workdir" -w /workdir --entrypoint make python:3 help
```
or
```
$  docker run --rm -ti -v "$PWD:/workdir" -w /workdir --entrypoint bash python:3 -c 'make init && make help'

Reply:
Of course `python:3` is contrived. The actual use case I've encountered is when using a CI service that itself runs in a container, e.g. AWS CodeBuild. To force-off the auto-init in that use case, we're now introducing a difference between how you run locally and how you run in CI. That kind of thing is confusing and frustrating in any project. Considering `BUILD_HARNESS_AUTO_INIT=true`, the intention is quite explicit, and of course the user can also disable that.

Anyway, that's the case I'm making. If you insist, I'll go ahead and disable it entirely for docker containers. (Of course, other container runtimes will still auto-init, because they don't generally match the `docker` pattern used in the grep.)

Reply:
Well, I guess you have to pick which "confusing and frustrating" behavior you prefer. Prior to your proposal, everything was consistent with regards to when you did or did not need to run `make init`. Your case for auto-init is that "people kept forgetting to run `make init`", and I am willing to allow your automation for "people" where they can supervise it. (There remains a confusing and frustrating change in behavior if you have the `build-harness` project checked out in a sibling directory, but that is not something we are going to address in this PR.)

I intentionally do not want to change the behavior of CI services because we have a lot of those installed and running and sometimes they do not call `make init` for security reasons. I do not want to create a security vulnerability based on the value of an environment variable. 

What might work and you might find preferable is to add a new pattern rule that invokes init first:

```make
%.init: init
	$(MAKE) $*
```

Then you can run `make readme.init` to auto-init before making readme. (I think it will work, but I can think of a few reasons it would not work, so test it out.)

Reply:
It's funny, before actually implementing this feature, I was making the same arguments to my team that you are making to me now. Then I started using it and I realized they were right. This feature provides a vastly superior experience. I'm confident that if you use it, you'll come around also.

Reply:
Made the requested change